internal/types/document.go

```diff
@@ -146,92 +179,85 @@ func (d *Document) Command() string {
 	return keys[0]
 }
 
-// add adds the value for the given key, returning error if that key is already present.
+// add adds the value for the given key.
+// If the key already exists, it will create a duplicate key.
 //
 // As a special case, _id always becomes the first key.
 func (d *Document) add(key string, value any) error {
-	if _, ok := d.m[key]; ok {
-		return fmt.Errorf("types.Document.add: key already present: %q", key)
-	}
-
-	// update keys slice
 	if key == "_id" {
 		// TODO check that value is not regex or array: https://github.com/FerretDB/FerretDB/issues/1235
 
 		// ensure that _id is the first field
-		d.keys = slices.Insert(d.keys, 0, key)
+		d.fields = slices.Insert(d.fields, 0, field{key, value})
 	} else {
-		d.keys = append(d.keys, key)
+		d.fields = append(d.fields, field{key, value})
 	}
 
-	d.m[key] = value
-
 	return nil
 }
 
 // Has returns true if the given key is present in the document.
 func (d *Document) Has(key string) bool {
-	_, ok := d.m[key]
-	return ok
+	for _, field := range d.fields {
+		if field.key == key {
+			return true
+		}
+	}
+
+	return false
 }
 
 // Get returns a value at the given key.
+// If there are duplicated keys in the document, it returns the first value.
 func (d *Document) Get(key string) (any, error) {
-	if value, ok := d.m[key]; ok {
-		return value, nil
+	for _, field := range d.fields {
+		if field.key == key {
+			return field.value, nil
+		}
 	}
 
 	return nil, fmt.Errorf("types.Document.Get: key not found: %q", key)
 }
 
 // Set sets the value for the given key, replacing any existing value.
+// If the key is duplicated, only first value will be replaced.
+// If the key doesn't exist, it will be set.
 //
 // As a special case, _id always becomes the first key.
-func (d *Document) Set(key string, value any) error {
-	// update keys slice
+func (d *Document) Set(key string, value any) {
 	if key == "_id" {
 		// TODO check that value is not regex or array: https://github.com/FerretDB/FerretDB/issues/1235
 
 		// ensure that _id is the first field
-		if i := slices.Index(d.keys, key); i >= 0 {
-			d.keys = slices.Delete(d.keys, i, i+1)
-		}
-		d.keys = slices.Insert(d.keys, 0, key)
-	} else {
-		if _, ok := d.m[key]; !ok {
-			d.keys = append(d.keys, key)
+		if i := slices.Index(d.Keys(), key); i >= 0 {
+			d.fields = slices.Delete(d.fields, i, i+1)
 		}
+		d.fields = slices.Insert(d.fields, 0, field{key, value})
+
+		return
 	}
 
-	if d.m == nil {
-		d.m = map[string]any{
-			key: value,
+	for i, f := range d.fields {
+		if f.key == key {
+			d.fields[i].value = value
+			return
 		}
-		return nil
 	}
 
-	d.m[key] = value
-	return nil
+	d.fields = append(d.fields, field{key, value})
 }
 
 // Remove the given key and return its value, or nil if the key does not exist.
+// If there are duplicate keys, only the first value is deleted and returned.
```

Comment:
Was this what we decided or what MongoDB does?

Reply:
This is a very internal thing, so it's up to us to decide.

Reply:
I'm happy with what we have, particular the documented comment :hugs:

Reply:
I think we should panic in that case

Reply:
fixed